pkg/workers/leader_election_mgr_test.go

```diff
@@ -185,3 +187,387 @@ func TestLeaderElectionManager_acquireLeaderLease(t *testing.T) {
+			fields: fields{
+				workers: []Worker{
+					&WorkerMock{
+						GetIDFunc: func() string {
+							return "01"
+						},
+						GetWorkerTypeFunc: func() string {
+							return "cluster"
+						},
+						IsRunningFunc: func() bool {
+							return false
+						},
+						GetSyncGroupFunc: func() *sync.WaitGroup {
+							return &sync.WaitGroup{}
+						},
+						StartFunc: func() {},
+					},
+				},
+				connectionFactory:                      db.NewMockConnectionFactory(nil),
+				leaderElectionReconcilerRepeatInterval: 15 * time.Second,
+				workerGrp:                              &sync.WaitGroup{},
+				tearDown:                               make(chan struct{}),
+			},
+			setupFn: func() {
+				mockEntry := map[string]interface{}{
+					"leader":  "01",
+					"expires": time.Now().Add(time.Hour),
+				}
+				mocket.Catcher.Reset().
+					NewMock().
+					WithQuery("SELECT * FROM leader_leases").
+					WithReply([]map[string]interface{}{mockEntry})
+				mocket.Catcher.NewMock().WithExecException().WithQueryException()
+			},
+		},
+	}
+	for _, testcase := range tests {
+		tt := testcase
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			t.Parallel()
+			tt.setupFn()
+			s := &LeaderElectionManager{
+				workers:                                tt.fields.workers,
+				connectionFactory:                      tt.fields.connectionFactory,
+				tearDown:                               tt.fields.tearDown,
+				leaderElectionReconcilerRepeatInterval: tt.fields.leaderElectionReconcilerRepeatInterval,
+				workerGrp:                              sync.WaitGroup{},
+			}
+			s.Start()
+			workerMock := tt.fields.workers[0].(*WorkerMock)
+			g.Expect(workerMock.calls.GetID).To(HaveLen(2))
+			g.Expect(workerMock.calls.GetWorkerType).To(HaveLen(1))
+			g.Expect(workerMock.calls.Start).To(HaveLen(1))
+		})
+	}
+}
+
+func TestLeaderElectionManager_Stop(t *testing.T) {
+	type fields struct {
+		tearDown  chan struct{}
+		workerGrp *sync.WaitGroup
+		workers   []Worker
+	}
+	tests := []struct {
+		name   string
+		fields fields
+	}{
+		{
+			name: "should return if teardown is nil",
+			fields: fields{
+				tearDown: nil,
+			},
+		},
+		{
+			name: "should stop the worker",
+			fields: fields{
+				tearDown:  make(chan struct{}),
+				workerGrp: &sync.WaitGroup{},
+				workers: []Worker{
+					&WorkerMock{
+						GetIDFunc: func() string {
+							return "02"
+						},
+						GetWorkerTypeFunc: func() string {
+							return "leader"
+						},
+						IsRunningFunc: func() bool {
+							return true
+						},
+						GetSyncGroupFunc: func() *sync.WaitGroup {
+							return &sync.WaitGroup{}
+						},
+						StopFunc: func() {},
+					},
+				},
+			},
+		},
+	}
+	for _, testcase := range tests {
+		tt := testcase
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			s := &LeaderElectionManager{
+				workers:   tt.fields.workers,
+				tearDown:  tt.fields.tearDown,
+				workerGrp: sync.WaitGroup{},
+			}
+			s.Stop()
```

Comment:
let's add some assertion after this call as well

e.g check if `.Stop()` was called etc

Reply:
Even though we will only confirm what we set in the mocks, `IsRunning()` can be checked like this:

```
	for _, testcase := range tests {
		tt := testcase
		t.Run(tt.name, func(t *testing.T) {
			g := NewWithT(t)
			t.Parallel()
			s := &LeaderElectionManager{
				workers:   tt.fields.workers,
				tearDown:  tt.fields.tearDown,
				workerGrp: sync.WaitGroup{},
			}
			if (len(s.workers)) > 0 {
				g.Expect(s.workers[0].IsRunning()).To(BeTrue())
			}
			s.Stop()
		})
	}
```

I find the workers stuff tricky to test and I am not familiar with this part of kfm code

Reply:
@pawelpaszki Not sure `g.Expect(s.workers[0].IsRunning()).To(BeTrue())`: that would simply invoke the mock `IsRunningFunc` method that we defined here: https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1172/files/13feb35a18ff857e20f695a7eaafe25f1aed2cc2#diff-f95dcc5a7c0365d81ab0b0499b14bb8222240cde002620e604fe4e02cb5f9d57R335

I believe the test here should check that the mock methods are correctly invoked. Maybe something like:
```go
g.Expect(s.workers[0].calls.Stop).To(HaveLen(1))
```

Reply:
makes sense @ziccardi